pkg/datarepair/checker/checker.go

```diff
@@ -127,8 +141,18 @@ func (checker *Checker) IdentifyInjuredSegments(ctx context.Context) (err error)
 
 					// reset durability stats for next iteration
 					checker.monStats = durabilityStats{}
+
+					err := checker.garbageService.Send(ctx, checker.pieceTracker, func() {
```

Comment:
Callbacks aren't very common in goland, however there are cases where they are the simpler solution so beforehand I want to say that this isn't about to ban callbacks, hence don't take this comment as callbacks should never be use in Go.

I see that the callback is only used for calls `gcWaitGroup.Done()` so my question is why the `garbageService.Send` has to be a asynchronous  spinning up a new goroutine and it isn't this function who spins up the goroutine?

Reply:
We originally had this function spinning up the goroutine. The problem had to do with setting `checker.pieceTracker = checker.garbageService.NewPieceTracker()` after sending. It is `garbageService.Send` that updates the `lastSendTime` which determines whether or not to create a noop piece tracker or a normal one.
If we create a goroutine in the checker, we must wait for everything to send before we can set the next piece tracker. So if the checker loop starts iterating before all the filters are sent, it will be using an out of date piece tracker. By having the goroutine inside `Send()`, we can update the `lastSendTime`, kick off a goroutine, and return, allowing us to set the new updated piecetracker before the next iteration of the checker.
I am not sure if there is a better way to design this, but the way it is currently seemed like the best solution of the options available (in my opinion)